###################################################################################################
#                             chIPseqOverChromosomes.R                                       #
#                                                                                                 #
#    In this script, we graphically depict the distribution of peaks from the Orc1cdc6 ChIP-seq   #
# over chromosomes. The input data were previously generated by the ChIP-seq pipeline described   #
# earlier. To execute this script, we recommend installing the Bioconductor libraries in Rstudio. #
#                                                                                                 #
# We suggest executing this script in Rstudio for enhanced visualization, facilitating specific   #
# adjustments.                                                                                    #
# Usage: saveCommand chIPseqOverChromosomes.R                                                #
#                                                                                                 #
# Copyleft (É”) 2023 by Thiago Andrade Franco                                                      #
#                    <thiago.franco.esib@esib.butantan.gov.br                                     #
#    01/09/2023: First version.                                                                   #
###################################################################################################
# First step download the necessary libraries.
packs <- c("ggplot2", "rio", "dplyr",
           "plyr","car", "ggrastr",
           "janitor","cowplot" )

# Checking the packages installed.
if(sum(as.numeric(!packs %in% installed.packages())) != 0){
  installer <- packs[!packs %in% installed.packages()]
  for(i in 1:length(installer)) {
    install.packages(installer, dependencies = T)
    break()}
  sapply(packs, require, character = T) 
} else {
  sapply(packs, require, character = T) 
}

# Biocondutor  packs.
if (!require("BiocManager",quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install("rtracklayer")
BiocManager::install("GenomicRanges")
BiocManager::install("ChIPseeker")
BiocManager::install("clusterProfiler")

library("GenomicRanges")
library(ChIPseeker)
library(clusterProfiler)
library(rtracklayer)

# Adjust the working directory.
setwd("~/Documents/myScripts/RScript/Genomic-Figures")

# Insert the raw data to data frame composition.
# Creating count values column.
# The tabyl() function from the janitor package is useful for creating tables of count values for distinct levels.
# of factor or character variables. The tabyl() function takes a data frame as the first argument followed by one.
# or two variables of interest.

# Import date to Orc1cdc6 ChIP-seq peaks.
orc1cdc6                          <-Import(file ="orc1cdc6.txt")
colnames(orc1cdc6)                <- c('Chromosome', 'Start', 'End')
orc1cdc6                          <-tabyl(orc1cdc6,Chromosome)
colnames(orc1cdc6)                <- c('Chromosome', 'Count', 'Percent')
orc1cdc6

# Import date to Orc1cdc6 ChIP-seq peaks  - chromosomes 1 to 9.
orc1cdc6C1to9                      <-Import(file ="orc1cdc6Chro1to9.txt")
colnames(orc1cdc6C1to9)            <- c('Chromosome', 'Start', 'End')
orc1cdc6C1to9                      <-tabyl(orc1cdc6C1to9,Chromosome)
colnames(orc1cdc6C1to9)            <- c('Chromosome', 'Count', 'Percent')
orc1cdc6C1to9

# Import date to Orc1cdc6 ChIP-seq peaks  - chromosomes 10 to 19.
orc1cdc6C10to19                    <-Import(file ="orc1cdc6Chro10to19.txt")
colnames(orc1cdc6C10to19)          <- c('Chromosome', 'Start', 'End')
orc1cdc6C10to19                      <-tabyl(orc1cdc6C10to19,Chromosome)
colnames(orc1cdc6C10to19)          <- c('Chromosome', 'Count', 'Percent')
orc1cdc6C10to19

# Import date to Orc1cdc6 ChIP-seq peaks  - chromosomes 20 to 29.
orc1cdc6C20to29                     <-Import(file ="orc1cdc6Chro20to29.txt")
colnames(orc1cdc6C20to29)           <- c('Chromosome', 'Start', 'End')
orc1cdc6C20to29                     <-tabyl(orc1cdc6C20to29,Chromosome)
colnames(orc1cdc6C20to29)           <- c('Chromosome', 'Count', 'Percent')
orc1cdc6C20to29

# Import date to Orc1cdc6 ChIP-seq peaks  - chromosomes 30 to 39.
orc1cdc6C30to39                     <-Import(file ="orc1cdc6Chro30to39.txt")
colnames(orc1cdc6C30to39)           <- c('Chromosome', 'Start', 'End')
orc1cdc6C30to39                     <-tabyl(orc1cdc6C30to39,Chromosome)
colnames(orc1cdc6C30to39)           <- c('Chromosome', 'Count', 'Percent')
orc1cdc6C30to39

# Import date to Orc1cdc6 ChIP-seq peaks  - chromosomes 40 and 41.
orc1cdc6C40to41                     <-Import(file ="orc1cdc6Chro40to41.txt")
colnames(orc1cdc6C40to41)           <- c('Chromosome', 'Start', 'End')
orc1cdc6C40to41                     <-tabyl(orc1cdc6C40to41,Chromosome)
colnames(orc1cdc6C40to41)           <- c('Chromosome', 'Count', 'Percent')
orc1cdc6C40to41

# Defining the factors for the data frame (database).
P_factor <- c(rep('Orc1Cdc6', dim(orc1cdc6)[1]))
P1_factor <- c(rep('Orc1Cdc6', dim(orc1cdc6C1to9)[1]))
P2_factor <- c(rep('Orc1Cdc6', dim(orc1cdc6C10to19)[1]))
P3_factor <- c(rep('Orc1Cdc6', dim(orc1cdc6C20to29)[1]))
P4_factor <- c(rep('Orc1Cdc6', dim(orc1cdc6C30to39)[1]))
P5_factor <- c(rep('Orc1Cdc6', dim(orc1cdc6C40to41)[1]))

# Column composition to Orc1Cdc6 ChIP-seq peaks.
orc1cdc6$V4                <- P_factor
colnames(orc1cdc6 )        <- c('Chromosome', 'Count','Percent','Origins')
orc1cdc6 
# Column composition to Orc1Cdc6 ChIP-seq peaks -chromosomes 1 to 9.
orc1cdc6C1to9$V4           <- P1_factor
colnames(orc1cdc6C1to9)    <- c('Chromosome', 'Count','Percent','Origins')
orc1cdc6C1to9
# Column composition to Orc1Cdc6 ChIP-seq peaks -chromosomes 10 to 19.
orc1cdc6C10to19$V4         <- P2_factor
colnames(orc1cdc6C10to19)  <- c('Chromosome', 'Count','Percent','Origins')
orc1cdc6C10to19
# Column composition to Orc1Cdc6 ChIP-seq peaks -chromosomes 20 to 29.
orc1cdc6C20to29$V4         <- P3_factor
colnames(orc1cdc6C20to29)  <- c('Chromosome', 'Count','Percent','Origins')
orc1cdc6C20to29
# Column composition to Orc1Cdc6 ChIP-seq peaks -chromosomes 30 to 39.
orc1cdc6C30to39$V4         <- P4_factor
colnames(orc1cdc6C30to39)  <- c('Chromosome', 'Count','Percent','Origins')
orc1cdc6C30to39
# Column composition to Orc1Cdc6 ChIP-seq peaks -chromosomes 40 and 41.
orc1cdc6C40to41$V4         <- P5_factor
colnames(orc1cdc6C40to41)  <- c('Chromosome', 'Count','Percent','Origins')
orc1cdc6C40to41

# Creating dataframe to Orc1Cdc6 ChIP-seq peaks.
df                        <-rbind(orc1cdc6)
colnames(df)              <- c('Chromosome', 'Count','Percent','Origins')
df
# Creating dataframe  to Orc1Cdc6 ChIP-seq peaks -chromosomes 1 to 9.
df1                        <-rbind(orc1cdc6C1to9)
colnames(df1)              <- c('Chromosome', 'Count','Percent','Origins')
df1
# Creating dataframe  to Orc1Cdc6 ChIP-seq peaks -chromosomes 10 to 19.
df2                        <-rbind(orc1cdc6C10to19)
colnames(df2)              <- c('Chromosome', 'Count','Percent','Origins')
df2
# Creating dataframe  to Orc1Cdc6 ChIP-seq peaks -chromosomes 20 to 29.
df3                        <-rbind(orc1cdc6C20to29)
colnames(df3)              <- c('Chromosome', 'Count','Percent','Origins')
df3
# Creating dataframe  to Orc1Cdc6 ChIP-seq peaks -chromosomes 30 to 39.
df4                        <-rbind(orc1cdc6C30to39)
colnames(df4)              <- c('Chromosome', 'Count','Percent','Origins')
df4
# Creating dataframe  to Orc1Cdc6 ChIP-seq peaks -chromosomes 40 to 41.
df5                        <-rbind(orc1cdc6C40to41)
colnames(df5)              <- c('Chromosome', 'Count','Percent','Origins')
df5

# Confirm your matrix is a data frame.
is.data.frame(df)
is.data.frame(df1)
is.data.frame(df2)
is.data.frame(df3)
is.data.frame(df4)
is.data.frame(df5)

# Ploting bar grapgh  with percentage.
## To plot Chip-seq data (full).
m <- ggplot(df, aes(x = Origins,y = Count, fill = Chromosome)) +
  geom_bar(stat="identity", position ="dodge") +
  theme_minimal()+
  theme(plot.title=element_text(size = 8, color="black")) +
  theme(axis.text = element_text(size = 8 , color ="black"),
        axis.title = element_text(size=8, color = "black"),
        legend.position = "bottom") +
  scale_fill_manual(values=c("#543005", "#8e0152", "#40004b", 
                             "#7f3b08", "#8c510a", "#de77ae", 
                             "#762a83", "#b35806", "#bf812d", 
                             "#f1b6da", "#9970ab","#e08214", 
                             "#dfc27d", "#fde0ef", "#c2a5cf",
                             "#fdb863", "#f6e8c3", "#f7f7f7", 
                             "#e7d4e8","#fee0b6", "#f5f5f5", 
                             "#e6f5d0", "#f7f7f7","#f7f7f7", 
                             "#c7eae5", "#c51b7d","#d9f0d3",
                             "#d8daeb", "#80cdc1", "#b8e186",
                             "#a6dba0","#b2abd2", "#35978f", 
                             "#7fbc41","#5aae61","#8073ac",
                             "#01665e", "#4d9221","#1b7837",
                             "#542788", "#003c30")) +
  labs(x ="", y = "Frequency")
m

# Ploting covplot to Orc1Cdc6 ChIP-seq peaks (full).
## library(rtracklayer)
peakOrc1cdc6 <- import.bed('Orc1Cdc6.bed')
peakOrc1cdc6

z <- covplot(peakOrc1cdc6, weightCol="score",lower= 1 , chrs =NULL, xlab = "Chromosome Size (bp)", ylab = "ChIp-Seq peaks frequency", title = NULL) +
  scale_x_continuous(breaks=c(0,500000,1000000,1500000,2000000), labels = c("0","500000","1000000","1500000","2000000"))
z

# We can plotting two different plots in 1 figure.
p <- plot_grid(m,z,  labels=c("A", "B"), ncol = 1, nrow = 2)
p
# Finally, we will save the plot
ggsave("orc1cdc6peaksChromoDistribuition-full.pdf", p,  height=8, width=12)

## To plot Chip-seq data chromosomes 1 to 9

m1 <- ggplot(df1, aes(x = Origins,y = Count, fill = Chromosome)) +
  geom_bar(stat="identity", position ="dodge") +
  theme_minimal()+
  theme(plot.title=element_text(size = 6, color="black")) +
  theme(axis.text = element_text(size = 6 , color ="black"),
        axis.title = element_text(size=6, color = "black"),
        legend.position = "bottom", 
        legend.text = element_text(size=6),
        legend.title = element_text(size=6)) +
  scale_fill_manual(values=c("#543005", "#8e0152", "#40004b",
                             "#7f3b08", "#8c510a", "#de77ae",
                             "#762a83", "#b35806", "#bf812d")) +
  labs(x ="", y = "Frequency")
m1

# Ploting covplot to Orc1Cdc6 ChIP-seq peaks (full).
z1 <- covplot(peakOrc1cdc6, weightCol="score",lower= 1 ,
chrs =c("TcChr1-S", "TcChr2-S","TcChr3-S", "TcChr4-S", "TcChr5-S","TcChr6-S","TcChr7-S","TcChr8-S","TcChr9-S"),
xlab = "Chromosome Size (bp)", ylab = "ChIp-Seq peaks frequency",
title = NULL) + 
  scale_x_continuous(breaks=c(0,100000,200000,300000,400000),
  labels = c("0","100000","200000","300000","400000"))
z1

# We can plotting two different plots in 1 figure.
p1 <- plot_grid(m1,z1,  labels=c("A", "B"), ncol = 2, nrow = 1)
p1
# Finally, we will save the plot
ggsave("Atlas/orc1cdc6peaksChromoDistribuition-1to9.pdf", p1,  height=8, width=11)

## To plot Chip-seq data chromosomes 10 to 19

m2 <- ggplot(df2, aes(x = Origins,y = Count, fill = Chromosome)) +
  geom_bar(stat="identity", position ="dodge") +
  theme_minimal()+
  theme(plot.title=element_text(size = 6, color="black")) +
  theme(axis.text = element_text(size = 6 , color ="black"),
        axis.title = element_text(size=6, color = "black"),
        legend.position = "bottom", 
        legend.text = element_text(size=6),
        legend.title = element_text(size=6)) +
  scale_fill_manual(values=c("#543005", "#8e0152", "#40004b",
                             "#7f3b08", "#8c510a", "#de77ae",
                             "#762a83", "#b35806", "#bf812d", "red")) +
  labs(x ="", y = "Frequency")
m2

# Ploting covplot to Orc1Cdc6 ChIP-seq peaks 1 to 9
z2 <- covplot(peakOrc1cdc6, weightCol="score",lower= 1 ,
              chrs =c("TcChr10-S", "TcChr11-S","TcChr12-S","TcChr13-S", "TcChr14-S", "TcChr15-S","TcChr16-S","TcChr17-S","TcChr18-S","TcChr19-S"),
              xlab = "Chromosome Size (bp)", ylab = "ChIp-Seq peaks frequency",
              title = NULL) + 
  scale_x_continuous(breaks=c(0, 100000,200000,300000,400000,500000,600000,700000),
                     labels = c("0", "100000","200000","300000","400000","500000","600000","700000" ))
z2

# We can plotting two different plots in 1 figure.
p2 <- plot_grid(m2,z2,  labels=c("A", "B"), ncol = 2, nrow = 1)
p2
# Finally, we will save the plot
ggsave("Atlas/orc1cdc6peaksChromoDistribuition-10to19.pdf", p2,  height=8, width=11)

## To plot Chip-seq data chromosomes 20 to 29

m3 <- ggplot(df3, aes(x = Origins,y = Count, fill = Chromosome)) +
  geom_bar(stat="identity", position ="dodge") +
  theme_minimal()+
  theme(plot.title=element_text(size = 6, color="black")) +
  theme(axis.text = element_text(size = 6 , color ="black"),
        axis.title = element_text(size=6, color = "black"),
        legend.position = "bottom", 
        legend.text = element_text(size=6),
        legend.title = element_text(size=6)) +
  scale_fill_manual(values=c("#543005", "#8e0152", "#40004b",
                             "#7f3b08", "#8c510a", "#de77ae",
                             "#762a83", "#b35806", "#bf812d", "red")) +
  labs(x ="", y = "Frequency")
m3

# Ploting covplot to Orc1Cdc6 ChIP-seq peaks 
z3 <- covplot(peakOrc1cdc6, weightCol="score",lower= 1 ,
              chrs =c("TcChr20-S", "TcChr21-S","TcChr22-S","TcChr23-S", "TcChr24-S", "TcChr25-S","TcChr26-S","TcChr27-S","TcChr28-S","TcChr29-S"),
              xlab = "Chromosome Size (bp)", ylab = "ChIp-Seq peaks frequency",
              title = NULL) + 
  scale_x_continuous(breaks=c(0, 200000,400000,600000,800000,1000000),
                     labels = c("0","200000","400000","600000","800000","1000000" ))
z3

# We can plotting two different plots in 1 figure.
p3 <- plot_grid(m3,z3,  labels=c("A", "B"), ncol = 2, nrow = 1)
p3
# Finally, we will save the plot
ggsave("Atlas/orc1cdc6peaksChromoDistribuition-20to29.pdf", p3,  height=8, width=11)

## To plot Chip-seq data chromosomes 30 to 39

m4 <- ggplot(df4, aes(x = Origins,y = Count, fill = Chromosome)) +
  geom_bar(stat="identity", position ="dodge") +
  theme_minimal()+
  theme(plot.title=element_text(size = 6, color="black")) +
  theme(axis.text = element_text(size = 6 , color ="black"),
        axis.title = element_text(size=6, color = "black"),
        legend.position = "bottom", 
        legend.text = element_text(size=6),
        legend.title = element_text(size=6)) +
  scale_fill_manual(values=c("#543005", "#8e0152", "#40004b",
                             "#7f3b08", "#8c510a", "#de77ae",
                             "#762a83", "#b35806", "#bf812d", "red")) +
  labs(x ="", y = "Frequency")
m4

# Ploting covplot to Orc1Cdc6 ChIP-seq peaks 
z4 <- covplot(peakOrc1cdc6, weightCol="score",lower= 1 ,
              chrs =c("TcChr30-S", "TcChr31-S","TcChr32-S",
                      "TcChr33-S", "TcChr34-S", "TcChr35-S",
                      "TcChr36-S","TcChr37-S","TcChr38-S","TcChr39-S"),
              xlab = "Chromosome Size (bp)", ylab = "ChIp-Seq peaks frequency",
              title = NULL) + 
  scale_x_continuous(breaks=c(0, 500000,1000000,1500000,2000000),
                              labels = c("0","500000","1000000","1500000","2000000" ))
z4

# We can plotting two different plots in 1 figure.
p4 <- plot_grid(m4,z4,  labels=c("A", "B"), ncol = 2, nrow = 1)
p4
# Finally, we will save the plot
ggsave("Atlas/orc1cdc6peaksChromoDistribuition-30to39.pdf", p4,  height=8, width=11)


## To plot Chip-seq data chromosomes 40 to 41

m5 <- ggplot(df5, aes(x = Origins,y = Count, fill = Chromosome)) +
  geom_bar(stat="identity", position ="dodge") +
  theme_minimal()+
  theme(plot.title=element_text(size = 6, color="black")) +
  theme(axis.text = element_text(size = 6 , color ="black"),
        axis.title = element_text(size=6, color = "black"),
        legend.position = "bottom", 
        legend.text = element_text(size=6),
        legend.title = element_text(size=6)) +
  scale_fill_manual(values=c("blue", "red")) +
  labs(x ="", y = "Frequency")
m5

# Ploting covplot to Orc1Cdc6 ChIP-seq peaks 
z5 <- covplot(peakOrc1cdc6, weightCol="score",lower= 1 ,
              chrs =c("TcChr40-S", "TcChr41-S"),
              xlab = "Chromosome Size (bp)", ylab = "ChIp-Seq peaks frequency",
              title = NULL) + 
  scale_x_continuous(breaks=c(0, 500000,1000000,1500000,2000000,3000000,4000000),
                     labels = c("0","500000","1000000","1500000","2000000","3000000","4000000" ))
z5

# We can plotting two different plots in 1 figure.
p5 <- plot_grid(m5,z5,  labels=c("A", "B"), ncol = 2, nrow = 1)
p5
# Finally, we will save the plot
ggsave("Atlas/orc1cdc6peaksChromoDistribuition-40to41.pdf", p5,  height=8, width=11)
